File: python/ahf_ai/providers/anthropic.py
"""Anthropic Messages API provider using raw httpx requests.

Talks directly to the Anthropic Messages API without the ``anthropic``
Python package, keeping the dependency tree minimal.
"""

from __future__ import annotations

import json
from typing import Any, AsyncIterator

import httpx

from ahf_ai.providers.base import BaseProvider

_DEFAULT_BASE_URL = "https://api.anthropic.com/v1"
_API_VERSION = "2023-06-01"

_MODEL_CONTEXT: dict[str, int] = {
    "claude-sonnet-4-20250514": 200_000,
    "claude-opus-4-20250514": 200_000,
    "claude-3-haiku": 200_000,
    "claude-3-sonnet": 200_000,
    "claude-3-opus": 200_000,
}


class AnthropicProvider(BaseProvider):
    """Provider for the Anthropic Messages API.

    Args:
        api_key: Anthropic API key.
        model: Model identifier (default ``claude-sonnet-4-20250514``).
        base_url: Override the API base URL.
        timeout: HTTP timeout in seconds.
    """

    def __init__(
        self,
        api_key: str,
        model: str = "claude-sonnet-4-20250514",
        base_url: str | None = None,
        timeout: float = 30.0,
    ) -> None:
        self._api_key = api_key
        self._model = model
        self._base_url = (base_url or _DEFAULT_BASE_URL).rstrip("/")
        self._timeout = timeout

    # -- BaseProvider implementation ------------------------------------------

    async def complete(
        self,
        messages: list[dict[str, str]],
        **kwargs: Any,
    ) -> str:
        system_text, user_messages = self._split_system(messages)
        payload: dict[str, Any] = {
            "model": self._model,
            "max_tokens": kwargs.pop("max_tokens", 1024),
            "messages": user_messages,
            **kwargs,
        }
        if system_text:
            payload["system"] = system_text

        async with httpx.AsyncClient(timeout=self._timeout) as client:
            resp = await client.post(
                f"{self._base_url}/messages",
                headers=self._headers(),
                json=payload,
            )
            resp.raise_for_status()
            data = resp.json()
        return data["content"][0]["text"]

    async def stream(
        self,
        messages: list[dict[str, str]],
        **kwargs: Any,
    ) -> AsyncIterator[str]:
        system_text, user_messages = self._split_system(messages)
        payload: dict[str, Any] = {
            "model": self._model,
            "max_tokens": kwargs.pop("max_tokens", 1024),
            "messages": user_messages,
            "stream": True,
            **kwargs,
        }
        if system_text:
            payload["system"] = system_text

        async with httpx.AsyncClient(timeout=self._timeout) as client:
            async with client.stream(
                "POST",
                f"{self._base_url}/messages",
                headers=self._headers(),
                json=payload,
            ) as resp:
                resp.raise_for_status()
                async for line in resp.aiter_lines():
                    if not line.startswith("data: "):
                        continue
                    data_str = line[len("data: "):]
                    chunk = json.loads(data_str)
                    if chunk.get("type") == "content_block_delta":
                        text = chunk["delta"].get("text", "")
                        if text:
                            yield text
                    elif chunk.get("type") == "message_stop":
                        break

    @property
    def model_name(self) -> str:
        return self._model

    @property
    def max_context_length(self) -> int:
        return _MODEL_CONTEXT.get(self._model, 200_000)

    # -- helpers --------------------------------------------------------------

    def _headers(self) -> dict[str, str]:
        return {
            "x-api-key": self._api_key,
            "anthropic-version": _API_VERSION,
            "Content-Type": "application/json",
        }

    @staticmethod
    def _split_system(
        messages: list[dict[str, str]],
    ) -> tuple[str | None, list[dict[str, str]]]:
        """Separate system messages from user/assistant messages.

        The Anthropic API takes ``system`` as a top-level parameter rather
        than as a message role.
        """
        system_parts: list[str] = []
        other: list[dict[str, str]] = []
        for msg in messages:
            if msg.get("role") == "system":
                system_parts.append(msg["content"])
            else:
                other.append(msg)
        system_text = "\n\n".join(system_parts) if system_parts else None
        return system_text, other

File: typescript/src/providers/anthropic.ts
/**
 * Anthropic Messages API provider implementation.
 *
 * Uses raw fetch against the /v1/messages endpoint with
 * the x-api-key header rather than Bearer auth.
 */

import type { Message } from "../types/index.js";
import { BaseProvider, type StreamChunk } from "./base.js";

export class AnthropicProvider extends BaseProvider {
  readonly modelName: string;
  readonly maxContextLength: number;
  private readonly apiKey: string;
  private readonly baseUrl: string;

  constructor(apiKey: string, model = "claude-sonnet-4-20250514", baseUrl = "https://api.anthropic.com") {
    super();
    this.apiKey = apiKey;
    this.modelName = model;
    this.baseUrl = baseUrl;
    this.maxContextLength = 200_000;
  }

  private headers(): Record<string, string> {
    return {
      "Content-Type": "application/json",
      "x-api-key": this.apiKey,
      "anthropic-version": "2023-06-01",
    };
  }

  async complete(messages: Message[]): Promise<{
    text: string;
    promptTokens: number;
    completionTokens: number;
    finishReason: string;
  }> {
    const systemMsg = messages.find((m) => m.role === "system");
    const nonSystem = messages.filter((m) => m.role !== "system");
    const body: Record<string, unknown> = {
      model: this.modelName,
      max_tokens: 4096,
      messages: nonSystem,
    };
    if (systemMsg) body["system"] = systemMsg.content;

    const res = await fetch(`${this.baseUrl}/v1/messages`, {
      method: "POST",
      headers: this.headers(),
      body: JSON.stringify(body),
    });
    if (!res.ok) throw await this.formatError(res);

    const data = await res.json() as {
      content: { text: string }[];
      stop_reason: string;
      usage: { input_tokens: number; output_tokens: number };
    };
    return {
      text: data.content.map((c) => c.text).join(""),
      promptTokens: data.usage.input_tokens,
      completionTokens: data.usage.output_tokens,
      finishReason: data.stop_reason === "end_turn" ? "stop" : data.stop_reason,
    };
  }

  async *stream(messages: Message[]): AsyncIterableIterator<StreamChunk> {
    const systemMsg = messages.find((m) => m.role === "system");
    const nonSystem = messages.filter((m) => m.role !== "system");
    const body: Record<string, unknown> = {
      model: this.modelName,
      max_tokens: 4096,
      messages: nonSystem,
      stream: true,
    };
    if (systemMsg) body["system"] = systemMsg.content;

    const res = await fetch(`${this.baseUrl}/v1/messages`, {
      method: "POST",
      headers: this.headers(),
      body: JSON.stringify(body),
    });
    if (!res.ok) throw await this.formatError(res);
    if (!res.body) throw new Error("Response body is null");

    const reader = res.body.getReader();
    const decoder = new TextDecoder();
    let buffer = "";

    while (true) {
      const { done, value } = await reader.read();
      if (done) break;
      buffer += decoder.decode(value, { stream: true });
      const lines = buffer.split("\n");
      buffer = lines.pop() ?? "";
      for (const line of lines) {
        if (!line.startsWith("data: ")) continue;
        const parsed = JSON.parse(line.slice(6)) as { type: string; delta?: { text?: string } };
        if (parsed.type === "content_block_delta" && parsed.delta?.text) {
          yield { text: parsed.delta.text, done: false };
        }
        if (parsed.type === "message_stop") {
          yield { text: "", done: true };
        }
      }
    }
  }
}
